lib/signet/sleuth.ex

```diff
@@ -50,77 +57,104 @@ defmodule Signet.Sleuth do
     end
   end
 
-  defp try_decode(query_res, selector) do
+  defp try_decode(query_res, selector, decode_structs) do
     try do
-      {:ok, ABI.decode(%ABI.FunctionSelector{types: selector.returns}, query_res)}
+      {:ok,
+       ABI.decode(
+         %ABI.FunctionSelector{types: selector.returns},
+         query_res,
+         decode_structs: decode_structs
+       )}
     rescue
       e ->
         {:error, "error decoding: #{inspect(e)}"}
     end
   end
 
-  defp encode_map(res, types, annotated, decode_binaries) do
-    Enum.map(Enum.zip(res, Enum.with_index(types)), fn {res, {type, i}} ->
-      encode_item(res, type, i, annotated, decode_binaries)
+  defp postprocess(results, named_types, opts) when is_list(results) and is_list(named_types) do
+    be_obvious = Keyword.get(opts, :be_obvious, false)
+
+    results
+    |> Enum.zip(named_types)
+    |> Enum.map(fn {it, t} -> {t.name, postprocess(it, t.type, opts)} end)
+    |> then(fn
+      processed_results when not be_obvious ->
+        case processed_results do
+          [] -> []
+          [{nil, result}] -> result
+          [{"", result}] -> result
+
+          [_more | _than_one] = processed_results ->
+            processed_results
+            |> Enum.with_index()
+            |> Enum.map(fn {{name, it}, i} ->
+              name =
+                if is_nil(name) or name == "" do
+                  "var#{i}"
+                else
+                  name
+                end
+
+              {name, it}
+            end)
+            |> Enum.into(%{})
+        end
```

Comment:
this drives me nuts. i kinda see why this ended up happening, but it still drives me nuts.